src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs

```diff
@@ -90,68 +78,54 @@ public void InvokeCompletionCallback(uint sniErrorCode)
         /// Allocate space for data
         /// </summary>
         /// <param name="capacity">Length of byte array to be allocated</param>
-        public void Allocate(int capacity)
+        public void Allocate(int capacity, bool reserveMuxHeader)
         {
-            if (_data != null && _data.Length < capacity)
+            SNIPacketFlags flags = reserveMuxHeader ? SNIPacketFlags.MuxHeaderReserved : SNIPacketFlags.None;
+            int headerCapacity = reserveMuxHeader ? SNISMUXHeader.HEADER_LENGTH : 0;
+            int totalCapacity = headerCapacity + capacity;
+            if (_data != null)
             {
-                if (_isBufferFromArrayPool)
+                if (_data.Length < totalCapacity)
                 {
-                    ArrayPool<byte>.Shared.Return(_data);
+                    Array.Clear(_data, 0, _header + _length);
+                    if ((_flags & SNIPacketFlags.ArrayFromPool) == SNIPacketFlags.ArrayFromPool)
+                    {
+                        ArrayPool<byte>.Shared.Return(_data, clearArray: false);
+                        _flags &= ~SNIPacketFlags.ArrayFromPool;
+                    }
+                    _data = null;
+                }
+                else
+                {
+                    // if the current array is big enough and rented keep it
+                    flags |= (_flags & SNIPacketFlags.ArrayFromPool);
                 }
-                _data = null;
             }
 
             if (_data == null)
             {
-                _data = ArrayPool<byte>.Shared.Rent(capacity);
-                _isBufferFromArrayPool = true;
+                _data = ArrayPool<byte>.Shared.Rent(totalCapacity);
+                flags |= SNIPacketFlags.ArrayFromPool; // set local not instance because it will be assigned after this block
```

Comment:
I don't see _data being allocated using a new. It will always be from the array pool. Any reason why we need this flag to signify that the SNIPacket is from the ArrayPool?

Reply:
Flags area all removed now. 

 have considered having a specific arraypool for packet buffers. The lifetime of the buffers we use isn't quite as temporary as other allocations. We know there will only be a small number of packets (roughly 2-3, in, out and attn) per connection active at any one time and in the common case of a single connection string packet size per application that would means the pool is small and only contains arrays of a specific size.